source/common/tracing/opentracing_driver_impl.cc

```diff
@@ -0,0 +1,168 @@
+public:
+  explicit OpenTracingHTTPHeadersReader(const Http::HeaderMap& request_headers)
+      : request_headers_(request_headers) {}
+
+  typedef std::function<opentracing::expected<void>(opentracing::string_view,
+                                                    opentracing::string_view)>
+      OpenTracingCb;
+
+  // opentracing::HTTPHeadersReader
+  opentracing::expected<opentracing::string_view>
+  LookupKey(opentracing::string_view key) const override {
+    const Http::HeaderEntry* entry;
+    Http::HeaderMap::Lookup lookup_result =
+        request_headers_.lookup(Http::LowerCaseString{key}, &entry);
+    switch (lookup_result) {
+    case Http::HeaderMap::Lookup::Found:
+      return opentracing::string_view{entry->value().c_str(), entry->value().size()};
+    case Http::HeaderMap::Lookup::NotFound:
+      return opentracing::make_unexpected(opentracing::key_not_found_error);
+    case Http::HeaderMap::Lookup::NotSupported:
+      return opentracing::make_unexpected(opentracing::lookup_key_not_supported_error);
+    default:
+      NOT_REACHED;
+    }
+  }
+
+  opentracing::expected<void> ForeachKey(OpenTracingCb f) const override {
+    request_headers_.iterate(header_map_callback, static_cast<void*>(&f));
+    return {};
+  }
+
+private:
+  const Http::HeaderMap& request_headers_;
+
+  static Http::HeaderMap::Iterate header_map_callback(const Http::HeaderEntry& header,
+                                                      void* context) {
+    OpenTracingCb* callback = static_cast<OpenTracingCb*>(context);
+    opentracing::string_view key{header.key().c_str(), header.key().size()};
+    opentracing::string_view value{header.value().c_str(), header.value().size()};
+    if ((*callback)(key, value)) {
+      return Http::HeaderMap::Iterate::Continue;
+    } else {
+      return Http::HeaderMap::Iterate::Break;
+    }
+  }
+};
+} // namespace
+
+OpenTracingSpan::OpenTracingSpan(bool use_single_header_propagation, bool use_tracer_propagation,
+                                 std::unique_ptr<opentracing::Span>&& span)
+    : use_single_header_propagation_(use_single_header_propagation),
+      use_tracer_propagation_(use_tracer_propagation), span_(std::move(span)) {}
+
+void OpenTracingSpan::finishSpan() { span_->Finish(); }
+
+void OpenTracingSpan::setOperation(const std::string& operation) {
+  span_->SetOperationName(operation);
+}
+
+void OpenTracingSpan::setTag(const std::string& name, const std::string& value) {
+  span_->SetTag(name, value);
+}
+
+void OpenTracingSpan::injectContext(Http::HeaderMap& request_headers) {
+  if (use_single_header_propagation_) {
+    // Inject the span context using Envoy's single-header format.
+    std::ostringstream oss;
+    const opentracing::expected<void> was_successful =
+        span_->tracer().Inject(span_->context(), oss);
+    if (!was_successful) {
+      ENVOY_LOG(warn, "Failed to inject span context: {}", was_successful.error().message());
```

Comment:
Same comments as below on error handling (sorry comments are coming out of order). In general, error handling should:
1) not log above debug if it can happen via user supplied input
2) stat
3) if it's something that just shouldn't happen, it should be an ASSERT or possibly RELEAS_ASSERT (please see error handling section in style guide)

Reply:
Changed to log at debug level, but still need to add a stat.